emc/dmri.py

```diff
@@ -0,0 +1,218 @@
+import attr
+import numpy as np
+import h5py
+import nibabel as nb
+from nitransforms.linear import Affine
+
+
+def _data_repr(value):
+    if value is None:
+        return "None"
+    return f"<{'x'.join(str(v) for v in value.shape)} ({value.dtype})>"
+
+
+@attr.s(slots=True)
+class DWI:
+    """Data representation structure for dMRI data."""
+
+    dataobj = attr.ib(default=None, repr=_data_repr)
+    """A numpy ndarray object for the data array, without *b=0* volumes."""
+    affine = attr.ib(default=None, repr=_data_repr)
+    """Best affine for RAS-to-voxel conversion of coordinates (NIfTI header)."""
+    brainmask = attr.ib(default=None, repr=_data_repr)
+    """A boolean ndarray object containing a corresponding brainmask."""
+    bzero = attr.ib(default=None, repr=_data_repr)
+    """
+    A *b=0* reference map, preferably obtained by some smart averaging.
+    If the :math:`B_0` fieldmap is set, this *b=0* reference map should also
+    be unwarped.
+    """
+    gradients = attr.ib(default=None, repr=_data_repr)
+    """A 2D numpy array of the gradient table in RAS+B format."""
+    em_affines = attr.ib(default=None)
+    """
+    List of :obj:`nitransforms.linear.Affine` objects that bring
+    DWIs (i.e., no b=0) into alignment.
+    """
+    fieldmap = attr.ib(default=None, repr=_data_repr)
+    """A 3D displacements field to unwarp susceptibility distortions."""
+    _filepath = attr.ib(default=Path(mkdtemp()) / "em_cache.h5", repr=False)
+    """A path to an HDF5 file to store the whole dataset."""
+
+    def __len__(self):
+        """Obtain the number of high-*b* orientations."""
+        return self.gradients.shape[-1]
+
+    def logo_split(self, index, with_b0=False):
+        """
+        Produce one fold of LOGO (leave-one-gradient-out).
+
+        Parameters
+        ----------
+        index : :obj:`int`
+            Index of the DWI orientation to be left out in this fold.
+
+        Return
+        ------
+        (train_data, train_gradients) : :obj:`tuple`
+            Training DWI and corresponding gradients.
+            Training data/gradients come **from the updated dataset**.
+        (test_data, test_gradients) :obj:`tuple`
+            Test 3D map (one DWI orientation) and corresponding b-vector/value.
+            The test data/gradient come **from the original dataset**.
+
+        """
+        if not Path(self._filepath).exists():
+            self.to_filename(self._filepath)
+
+        # read original DWI data & b-vector
+        with h5py.File(self._filepath, "r") as in_file:
+            root = in_file["/0"]
+            dwframe = np.asanyarray(root["dataobj"][..., index])
+            bframe = np.asanyarray(root["gradients"][..., index])
+
+        # if the size of the mask does not match data, cache is stale
+        mask = np.zeros(len(self), dtype=bool)
+        mask[index] = True
+
+        train_data = self.dataobj[..., ~mask]
+        train_gradients = self.gradients[..., ~mask]
+
```

Comment:
So I think it would still be good to have some mechanism for further filtering which gradient samples are included for model training and testing upon each iteration. On the one hand, there are those samples that are too close in proximity along the sphere to those which we are trying to predict and therefore risk unfavorably biasing the predictions. But even if that's still a matter of debate, there may simply be outlier volumes or gradients that we wish to manually exclude. For that reason alone, I think it would still be beneficial to assume some additional level of control over which samples get used, no?

On that note, I do think that a threshold for filtering out training samples based on q-space proximity should itself be adaptive since I'd assume that the extent of this bias will vary as a function of model choices and sampling scheme (e.g. we should also think about what might happen here in the half-sphere case and how best to accommodate...).

But for now, adding some functionality here that does something like the following might be a good starting point:
```
def train_gradient_filter(
    test_gradients, train_gradients, cutoff=2.0
):
    train_bvec = np.asanyarray(train_gradients[:, 0:3])
    train_bval = np.asanyarray(train_gradients[:, 3])
    test_bvec = np.asanyarray(test_gradients[:, 0:3])
    test_bval = np.asanyarray(test_gradients[:, 3])
    
    min_bval = min(min(train_bval), test_bval)
    all_qvals = np.sqrt(train_bval - min_bval)
    prediction_qval = np.sqrt(test_bval - min_bval)

    # Convert q values to percent of maximum qval
    max_qval = max(max(all_qvals), prediction_qval)
    all_qvals_scaled = all_qvals / max_qval * 100
    scaled_qvecs = train_bvec * all_qvals_scaled[:, np.newaxis]
    scaled_prediction_qvec = test_bvec * (prediction_qval / max_qval * 100)
    
    # Calculate the distance between the training qvecs and the testing qvec
    gradient_mask = (
        np.linalg.norm(scaled_qvecs - scaled_prediction_qvec, axis=1
                       ) > cutoff
    ) * (np.linalg.norm(scaled_qvecs + scaled_prediction_qvec, axis=1
                        ) > cutoff)

    return gradient_mask
    
gradient_mask = train_gradient_filter(test_gradients, train_gradients)
train_data = self.dataobj[..., ~gradient_mask]
train_gradients = self.gradients[..., ~gradient_mask]
```
And then maybe also adding an `exclusion_indices` argument as well for manually dropping training samples if needed.

WDYT @oesteban ?

Reply:
> therefore risk unfavorably biasing the predictions.

It is still unclear to me why the bias would be unfavorable in this case

> there may simply be outlier volumes or gradients that we wish to manually exclude.

This is a chicken-and-egg problem. How do you decide what volumes should be excluded? Perhaps they'd be acceptable after head motion? I prefer to make the implementation as agnostic as possible of the rest of the pipeline. This is not to say that we should not look into this issue, but my opinion is that we can only investigate this outlier filtering after we have successfully tried the code on a lot of data. We need to measure how much it improves w.r.t. a baseline. This PR tries to set such a baseline.

Instead of thinking about solutions to hypothetical problems, let's hypothesize about the most relevant problems we expect first, then plan for a test that would evaluate solutions to the problem, and finally once we have both hypothesis and test, make decisions to improve performance.

Finally, and in particular about the code you propose:
- I would not implement this in the DWI object because this is not partitioning the dataset, it is partioning the partition with some additional criteria. IMHO, the responsibility of this falls more on the estimator object side.
- Let's keep this code around and test it when we have a functional prototype (i.e., a baseline as I called it before). This first implementation only looked for setting up the foundations.

Reply:
For instance, one problem I do see coming at us:

Say you have 12 b=1000, 12 b=2000 and 12 b=3000. Eddy currents for the two outer shells will be much larger than those of the b=1000. Because the dataset is already pretty biased towards the outer shells, when aligning one orientation of the b=1000 shell we might actually "overcorrect" for eddy. So in this case, you might want to completely dismiss the largest b (instead of close orientations). WDYT @arokem ?